File: src/main/java/com/example/tutorial/dsa/medium/stackqueue/QueueUsingStacks.java
package com.example.tutorial.dsa.medium.stackqueue;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * QueueUsingStacks
 * ----------------------------------
 * <p>This program implements a FIFO queue using two stacks.
 * The core problem solved here is Implement Queue using Stacks (LeetCode #232).
 *
 * <p><b>Problem Statement:</b>
 * Implement a first in first out (FIFO) queue using only two stacks. The queue
 * should support all functions: push, pop, peek, and empty.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Process offer redemption requests in order received</li>
 *   <li>Handle customer service tickets FIFO when only stack-based storage available</li>
 *   <li>Manage batch processing jobs in order</li>
 * </ul>
 *
 * <p><b>Operations:</b>
 * <ul>
 *   <li>push(x) - Push element to back of queue</li>
 *   <li>pop() - Remove element from front of queue</li>
 *   <li>peek() - Get front element</li>
 *   <li>empty() - Check if queue is empty</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon, Microsoft, Apple, Bloomberg
 *
 * @see <a href="https://leetcode.com/problems/implement-queue-using-stacks/">LeetCode 232</a>
 */
@Component
public class QueueUsingStacks implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(QueueUsingStacks.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== QueueUsingStacks: Queue Implementation Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate with customer service tickets
    System.out.println("--- Processing Service Tickets (FIFO) ---\n");
    MyQueue ticketQueue = new MyQueue();

    String[] tickets = {"Ticket-101", "Ticket-102", "Ticket-103", "Ticket-104"};
    System.out.println("Submitting tickets:");
    for (String ticket : tickets) {
      ticketQueue.push(ticket.hashCode());
      System.out.println("  Submitted: " + ticket);
    }

    System.out.println("\nProcessing tickets (FIFO order):");
    while (!ticketQueue.empty()) {
      System.out.println("  Processing ticket with hash: " + ticketQueue.pop());
    }

    // Standard LeetCode example
    System.out.println("\n--- LeetCode Example ---\n");
    MyQueue queue = new MyQueue();
    queue.push(1);
    System.out.println("push(1)");
    queue.push(2);
    System.out.println("push(2)");
    System.out.println("peek() → " + queue.peek());   // 1
    System.out.println("pop() → " + queue.pop());     // 1
    System.out.println("empty() → " + queue.empty()); // false
  }

  /**
   * MyQueue - Queue implemented using two stacks.
   *
   * <p><b>DESIGN (Two-Stack Reversal):</b>
   * <ol>
   *   <li>inputStack: receives all push operations</li>
   *   <li>outputStack: used for pop/peek operations</li>
   *   <li>When outputStack is empty and we need pop/peek:
   *       transfer all elements from inputStack to outputStack</li>
   *   <li>Transfer reverses order, giving us FIFO behavior!</li>
   * </ol>
   *
   * <p><b>Why does this work?</b>
   * <br>Stack is LIFO (Last In First Out). By moving elements from one stack to another,
   * we reverse the order. So the first element pushed into inputStack becomes the top
   * of outputStack - exactly what we need for a queue!
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * push(1): inputStack=[1], outputStack=[]
   * push(2): inputStack=[1,2], outputStack=[]
   * push(3): inputStack=[1,2,3], outputStack=[]
   *
   * pop(): outputStack empty, transfer!
   *        inputStack=[], outputStack=[3,2,1]
   *        pop from outputStack → returns 1
   *        outputStack=[3,2]
   *
   * push(4): inputStack=[4], outputStack=[3,2]
   *
   * pop(): outputStack not empty, pop directly → returns 2
   *        outputStack=[3]
   * </pre>
   *
   * <p><b>Time Complexity:</b>
   * <br>- Push: O(1) always
   * <br>- Pop/Peek: Amortized O(1) - each element transferred at most once
   * <br><i>Like moving papers between two trays - each paper is moved at most twice
   * in its lifetime (once in, once to output tray). Average cost per operation = constant.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>Total elements stored across both stacks = n.
   * <br><i>Like having two in-trays - papers are in one or the other, never both.</i>
   */
  public static class MyQueue {
    // Stack for push operations.
    private final Deque<Integer> inputStack;

    // Stack for pop/peek operations (reversed order).
    private final Deque<Integer> outputStack;

    public MyQueue() {
      inputStack = new ArrayDeque<>();
      outputStack = new ArrayDeque<>();
    }

    /**
     * Pushes element to the back of queue.
     *
     * @param x the element to push
     */
    public void push(int x) {
      inputStack.push(x);
    }

    /**
     * Removes and returns the front element.
     *
     * @return the front element
     */
    public int pop() {
      // Ensure outputStack has elements.
      transferIfNeeded();
      return outputStack.pop();
    }

    /**
     * Gets the front element without removing it.
     *
     * @return the front element
     */
    public int peek() {
      transferIfNeeded();
      return outputStack.peek();
    }

    /**
     * Checks if queue is empty.
     *
     * @return true if empty
     */
    public boolean empty() {
      return inputStack.isEmpty() && outputStack.isEmpty();
    }

    /**
     * Transfers elements from inputStack to outputStack if outputStack is empty.
     * This reverses the order, converting LIFO to FIFO.
     */
    private void transferIfNeeded() {
      if (outputStack.isEmpty()) {
        while (!inputStack.isEmpty()) {
          outputStack.push(inputStack.pop());
        }
      }
    }
  }
}

File: src/main/python/com/example/tutorial/dsa/medium/stackqueue/queue_using_stacks.py
"""
QueueUsingStacks
----------------------------------
This program implements a FIFO queue using two stacks.
The core problem solved here is Implement Queue using Stacks (LeetCode #232).

Problem Statement:
    Implement a first in first out (FIFO) queue using only two stacks. The queue
    should support all functions: push, pop, peek, and empty.

Real UseCase:
    In a credit card offers system:
    - Process offer redemption requests in order received
    - Handle customer service tickets FIFO when only stack-based storage available
    - Manage batch processing jobs in order

Operations:
    - push(x) - Push element to back of queue
    - pop() - Remove element from front of queue
    - peek() - Get front element
    - empty() - Check if queue is empty

Company Tags: Amazon, Microsoft, Apple, Bloomberg

See: https://leetcode.com/problems/implement-queue-using-stacks/
"""

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


class MyQueue:
    """
    MyQueue - Queue implemented using two stacks.

    DESIGN (Two-Stack Reversal):
        1. input_stack: receives all push operations
        2. output_stack: used for pop/peek operations
        3. When output_stack is empty and we need pop/peek:
           transfer all elements from input_stack to output_stack
        4. Transfer reverses order, giving us FIFO behavior!

    Why does this work?
        Stack is LIFO (Last In First Out). By moving elements from one stack to another,
        we reverse the order. So the first element pushed into input_stack becomes the top
        of output_stack - exactly what we need for a queue!

    Example Walkthrough:
        push(1): input_stack=[1], output_stack=[]
        push(2): input_stack=[1,2], output_stack=[]
        push(3): input_stack=[1,2,3], output_stack=[]

        pop(): output_stack empty, transfer!
               input_stack=[], output_stack=[3,2,1]
               pop from output_stack -> returns 1
               output_stack=[3,2]

        push(4): input_stack=[4], output_stack=[3,2]

        pop(): output_stack not empty, pop directly -> returns 2
               output_stack=[3]

    Time Complexity:
        - Push: O(1) always
        - Pop/Peek: Amortized O(1) - each element transferred at most once
        Like moving papers between two trays - each paper is moved at most twice
        in its lifetime (once in, once to output tray). Average cost per operation = constant.

    Space Complexity: O(n)
        Total elements stored across both stacks = n.
        Like having two in-trays - papers are in one or the other, never both.
    """

    def __init__(self):
        """Initialize the queue."""
        # Stack for push operations.
        self.input_stack = []

        # Stack for pop/peek operations (reversed order).
        self.output_stack = []

    def push(self, x: int) -> None:
        """
        Pushes element to the back of queue.

        Args:
            x: The element to push.
        """
        self.input_stack.append(x)

    def pop(self) -> int:
        """
        Removes and returns the front element.

        Returns:
            The front element.
        """
        # Ensure output_stack has elements.
        self._transfer_if_needed()
        return self.output_stack.pop()

    def peek(self) -> int:
        """
        Gets the front element without removing it.

        Returns:
            The front element.
        """
        self._transfer_if_needed()
        return self.output_stack[-1]

    def empty(self) -> bool:
        """
        Checks if queue is empty.

        Returns:
            True if empty.
        """
        return not self.input_stack and not self.output_stack

    def _transfer_if_needed(self) -> None:
        """
        Transfers elements from input_stack to output_stack if output_stack is empty.
        This reverses the order, converting LIFO to FIFO.
        """
        if not self.output_stack:
            while self.input_stack:
                self.output_stack.append(self.input_stack.pop())


def main():
    """Main function to demonstrate the QueueUsingStacks."""
    print("=== QueueUsingStacks: Queue Implementation Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate with customer service tickets
    print("--- Processing Service Tickets (FIFO) ---\n")
    ticket_queue = MyQueue()

    tickets = ["Ticket-101", "Ticket-102", "Ticket-103", "Ticket-104"]
    print("Submitting tickets:")
    for ticket in tickets:
        ticket_queue.push(hash(ticket))
        print(f"  Submitted: {ticket}")

    print("\nProcessing tickets (FIFO order):")
    while not ticket_queue.empty():
        print(f"  Processing ticket with hash: {ticket_queue.pop()}")

    # Standard LeetCode example
    print("\n--- LeetCode Example ---\n")
    queue = MyQueue()
    queue.push(1)
    print("push(1)")
    queue.push(2)
    print("push(2)")
    print(f"peek() -> {queue.peek()}")    # 1
    print(f"pop() -> {queue.pop()}")      # 1
    print(f"empty() -> {queue.empty()}")  # False


if __name__ == "__main__":
    main()
